### PR TITLE
kvdb: no overlay

### LINKED ISSUE #310: kvdb-rocksdb with no overlay cache
Currently, substrate is not using it, and probably for a good reason, since caching something that RocksDB caches (or is able to) on it's own is generally not a good idea.

### PR DESCRIPTION
**Pull request by ordian**
Fixes #310.

### CODE REVIEW

File: kvdb/src/lib.rs [hunk 1]
```diff
@@ -93,20 +93,11 @@ impl DBTransaction {
 
 /// Generic key-value database.
 ///
-/// This makes a distinction between "buffered" and "flushed" values. Values which have been
-/// written can always be read, but may be present in an in-memory buffer. Values which have
-/// been flushed have been moved to backing storage, like a RocksDB instance. There are certain
-/// operations which are only guaranteed to operate on flushed data and not buffered,
-/// although implementations may differ in this regard.
```
Comment (ordian) [inline kvdb/src/lib.rs]: > Anyone who relies on this semantics shouldn't have done it on the first place, it is not written anywhere (is it?)

@NikVolf ^^^ :P

Comment (NikVolf) [reply kvdb/src/lib.rs]: This is probably about iterators? :)

**Comment by NikVolf** [comment]
I wonder if we can get rid of locks on `self.db`, but it's a minor issue
probably required for safe closing/upgrading

**Comment by ordian** [comment]
> I wonder if we can get rid of locks on `self.db`, but it's a minor issue
> probably required for safe closing/upgrading

I considered using https://docs.rs/arc-swap/0.4.4/arc_swap/, but `read` locks are usually cheap enough, also not sure if holding a read lock during iteration to prevent the db from closing is a good idea in the end.

File: kvdb-rocksdb/src/lib.rs [hunk 2]
```diff
@@ -532,84 +443,55 @@ impl Database {
...
 	}
 
-	/// Get value by partial key. Prefix size should match configured prefix size. Only searches flushed values.
-	// TODO: support prefix seek for unflushed data
+	/// Get value by partial key. Prefix size should match configured prefix size.
 	pub fn get_by_prefix(&self, col: u32, prefix: &[u8]) -> Option<Box<[u8]>> {
 		self.iter_from_prefix(col, prefix).next().map(|(_, v)| v)
 	}
 
-	/// Get database iterator for flushed data.
```
Comment (NikVolf) [inline kvdb-rocksdb/src/lib.rs]: it was probably lying before?
since it was interleaved with actual data?

Comment (ordian) [reply kvdb-rocksdb/src/lib.rs]: partially, I guess
in order to support non-flushed data, it would have to take `flushing` into account (see e.g. `get`)

**Comment by ordian** [comment]
> > I wonder if we can get rid of locks on `self.db`, but it's a minor issue
> > probably required for safe closing/upgrading
> 
> I considered using https://docs.rs/arc-swap/0.4.4/arc_swap/, but `read` locks are usually cheap enough, also not sure if holding a read lock during iteration to prevent the db from closing is a good idea in the end.

Let's move this conversation to #314.

**Comment by NikVolf** [comment]
Great stuff!

File: kvdb/src/lib.rs [hunk 3]
```diff
@@ -118,25 +109,16 @@ pub trait KeyValueDB: Sync + Send + parity_util_mem::MallocSizeOf {
 	/// Get a value by key.
 	fn get(&self, col: u32, key: &[u8]) -> io::Result<Option<DBValue>>;
 
-	/// Get a value by partial key. Only works for flushed data.
+	/// Get a value by partial key. Only works for the data.
```
[the remark this answers is not in the document]
Comment (dvdplm) [reply kvdb/src/lib.rs]: Is it first or any random value?

File: kvdb/src/lib.rs [hunk 4]
```diff
@@ -118,25 +109,16 @@ pub trait KeyValueDB: Sync + Send + parity_util_mem::MallocSizeOf {
...
-	fn flush(&self) -> io::Result<()>;
+	fn write(&self, transaction: DBTransaction) -> io::Result<()>;
 
-	/// Iterate over flushed data for a given column.
+	/// Iterate over the data for a given column.
 	fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a>;
 
-	/// Iterate over flushed data for a given column, starting from a given prefix.
+	/// Iterate over the data for a given column, starting from a given prefix.
 	fn iter_from_prefix<'a>(
```
Comment (ordian) [inline kvdb/src/lib.rs]: maybe we want to finally rename it, but let's do it in a separate PR

File: kvdb-rocksdb/src/lib.rs [hunk 5]
```diff
@@ -532,84 +443,55 @@ impl Database {
...
+	/// Get database iterator for the data.
 	/// Will hold a lock until the iterator is dropped
 	/// preventing the database from being closed.
```
[the remark this answers is not in the document]
Comment (ordian) [reply kvdb-rocksdb/src/lib.rs]: this is a good question, but I consider it to be part of the #314

File: kvdb/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ordian) [reply kvdb/src/lib.rs]: First in the current implementation and *I think* we're relying on that, but maybe I'm wrong

**Comment by dvdplm** [comment]
So, any progress on this? I'd love to see some benchmarks on both eth and substrate before moving ahead.

**Comment by ordian** [comment]
I did import benchmarks on eth side https://github.com/paritytech/parity-common/issues/310#issuecomment-572657722, but we currently lack some read-heavy benches like RPC load etc. Would be nice if someone could bench it in substrate as well.

**Comment by ordian** [comment]
Getting back to this, since substrate doesn't use `write_buffered` method at all, I don't expect this PR to have any effect on this. For eth, it's only used in warp-sync, not regular import, so even if it will slow down warp-sync import by 5%, I think it's fine.

@dvdplm let me know what you think

**Comment by dvdplm** [comment]
> For eth, `write_buffered` only used in warp-sync, not regular import, so even if it will slow down warp-sync import by 5%, I think it's fine.

`rg "write_buffered\(" | wc -l` shows 19 calls in eth. Many of them look like they are not perf sensitive at all, but [`commit_block()`](https://github.com/openethereum/openethereum/blob/87fd914c942bb496209b763db08f8a3b1843ed45/ethcore/src/client/client.rs#L488) and [`feed()`](https://github.com/openethereum/openethereum/blob/3ccfe735aa579f89a1f8de192a7625084a6f530f/ethcore/snapshot/src/lib.rs#L399) are pretty important calls to keep fast. A 5% pessimization is probably not ok there. :/

I really like this PR and I want it to land, but I still want to be convinced we have hard data to support the decision.

**Comment by ordian** [comment]
See https://github.com/paritytech/parity-common/issues/310#issuecomment-573629718 about eth usage of it.

**Comment by dvdplm** [comment]
> See [#310 (comment)](https://github.com/paritytech/parity-common/issues/310#issuecomment-573629718) about eth usage of it.

Oops, apologies. K, fine let's do this then. :)
